### PR TITLE
Fix make_simplified_union interaction with Any

### LINKED ISSUE #2712: Union of a class with Any base and None is sometimes None
Without `--strict-optional`, the second `reveal_type` produces an incorrect result:

```
from typing import Any, Union, TypeVar

class C(Any):
    pass

T = TypeVar('T')
S = TypeVar('S')

def f(x: T, y: S) -> Union[T, S]:
    pass

reveal_type(f(C(), None))  # t.C
reveal_type(f(None, C()))  # builtins.None   <--- wrong
```

With `--strict-optional`, both of them are wrong:

```
...   # as above

reveal_type(f(C(), None))  # builtins.None   <--- wrong
reveal_type(f(None, C()))  # builtins.None   <--- wrong
```

This seems to cause problems in real-world code together with python/typeshed@6008b9d, so this is high priority.

### PR DESCRIPTION
**Pull request by JukkaL**
DO NOT MERGE YET -- this causes errors in internal Dropbox
repos.

(This is an updated version of #2197 by ddfisher.)

Fixes #2712.

make_simplified_union had two incorrect interactions with Any that this
fixes:
1) Any unions containing Any were simplified to Any, which is incorrect.
2) Any classes that inherited from Any would be removed from the union
by the subclass simplification (because subclasses of Any are
considered subclasses of all other classes due to subtype/compatible
with confusion).

Note that `Union`s call make_union and not make_simplified_union, so
this doesn't change their behavior directly (unless they interact with
something else which causes them to be simplified, like being part of
an `Optional`).

### CODE REVIEW

**Comment by JukkaL** [comment]
I'm thinking about first merging a simplified version of this that is just enough to fix #2712.

File: mypy/checkmember.py [hunk 1]
```diff
@@ -107,11 +107,17 @@ def analyze_member_access(name: str,
...
+        old_num_messages = msg.num_messages()
         results = [analyze_member_access(name, subtype, node, is_lvalue, is_super,
                                          is_operator, builtin_type, not_ready_callback, msg,
                                          original_type=original_type, chk=chk)
                    for subtype in typ.items]
         msg.disable_type_names -= 1
+        if msg.num_messages() != old_num_messages and any(isinstance(t, AnyType)
+                                                          for t in results):
+            # If there was an error, return AnyType to avoid generating multiple messages for the
+            # same error.
```
Comment (ddfisher) [inline mypy/checkmember.py]: Was this not a problem before?

File: mypy/meet.py [hunk 2]
```diff
@@ -25,9 +25,19 @@ def meet_types(s: Type, t: Type) -> Type:
...
+    """Return the type s "narrowed down" to t.
+
+    Note that this is not symmetric with respect to s and t.
+
+    TODO: Explain what this does in more detail and how this is
+          different from meet_types.
+    """
     if s == t:
         return s
+    if isinstance(t, AnyType):
```
Comment (ddfisher) [inline mypy/meet.py]: Why this change?  What effect does it have overall?

Comment (JukkaL) [inline mypy/meet.py]: Without this, `testUnionAnyIsInstance` will fail -- the type in the else block will be `Union[int, Any]`, though it should be `Any`. This makes is possible to narrow down from `Union[int, Any]` to `Any`, and this is the only place where this code is used. Previously this worked (probably accidentally) because `Union[int, Any]` got simplified to just `Any`. This won't have any other effect, as far I can see.

File: mypy/checkmember.py (hunk 1 above)
Comment (JukkaL) [inline mypy/checkmember.py]: This fixes a regression in `testUnionAndBinaryOperation` (and another test case whose name I can't remember right now). Previously, if some of the results was `Any`, the result got simplified into `Any`. Without this change, we'd preserve the non-`Any` types and this would sometimes produce a duplicate, similar-looking error message about the same binary operation, which would be a regression. Now the code falls back to the old behavior in these cases, and since this only happens when we have already generated a message, it doesn't affect correctness, just the specific messages generated for type errors involving union types.

**Comment by JukkaL** [comment]
I now feel that this approach is not quite right. A better way to simplify union may be use a subtype check that doesn't treat `Any` specially. We already have `is_proper_subtype` but it's very incomplete and probably not sufficient right now. I'll write more about this later.

**Comment by JukkaL** [comment]
Closing in favor of #3025.
